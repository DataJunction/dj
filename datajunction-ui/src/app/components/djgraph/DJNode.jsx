import React, { memo } from 'react';
import { Handle, Position } from 'reactflow';
import DJClientContext from '../../providers/djclient';
import { useContext, useEffect, useState } from 'react';

function renderBasedOnDJNodeType(param) {
  switch (param) {
    case 'source':
      return { backgroundColor: '#7EB46150', color: '#7EB461' };
    case 'transform':
      return { backgroundColor: '#6DAAA750', color: '#6DAAA7' };
    case 'dimension':
      return { backgroundColor: '#CF7D2950', color: '#CF7D29' };
    case 'metric':
      return { backgroundColor: '#A27E8650', color: '#A27E86' };
    case 'cube':
      return { backgroundColor: '#C2180750', color: '#C21807' };
    default:
      return {};
  }
}

function capitalize(string) {
  return string.charAt(0).toUpperCase() + string.slice(1);
}

const Collapse = ({ collapsed, text, children }) => {
  const [isCollapsed, setIsCollapsed] = React.useState(collapsed);

  return (
    <>
      <div className="collapse">
        <button
          className="collapse-button"
          onClick={() => setIsCollapsed(!isCollapsed)}
        >
          {isCollapsed ? '\u25B6 Show' : '\u25BC Hide'} {text}
        </button>
        <div
          className={`collapse-content ${
            isCollapsed ? 'collapsed' : 'expanded'
          }`}
          aria-expanded={isCollapsed}
        >
          {children}
        </div>
      </div>
    </>
  );
};

export function DJNode({ id, data }) {
  const columnsRenderer = data =>
    data.column_names.map(col => (
      <tr>
        <td>
          {data.primary_key.includes(col.name) ? (
            <b>{col.name} (PK)</b>
          ) : (
            <>{col.name}</>
          )}
        </td>
        <td style={{ textAlign: 'right' }}>{col.type}</td>
      </tr>
    ));

  const [dimensions, setDimensions] = useState([]);
  const djClient = useContext(DJClientContext).DataJunctionAPI;
  useEffect(() => {
    if (data.type === 'metric') {
      async function getDimensions() {
        try {
          const metricData = await djClient.metric(data.name);
          setDimensions(metricData.dimensions);
        } catch (err) {
          console.log(err);
        }
      }
      getDimensions();
    }
  }, [data, djClient]);

  const dimensionsRenderer = dimensions =>
    dimensions.map(dim => (
      <tr>
        <a href={`/nodes/${dim.name.substring(0, dim.name.lastIndexOf('.'))}`}>
          <td>{dim.name}</td>
        </a>
      </tr>
    ));

  return (
    <>
      <Handle
        type="target"
        position={Position.Left}
        style={{ backgroundColor: '#ccc' }}
      />
      <div className="dj-node__full" style={renderBasedOnDJNodeType(data.type)}>
        <div className="dj-node__header">
          <div className="serif">
            {data.name
              .split('.')
              .slice(0, data.name.split('.').length - 1)
              .join(' \u25B6 ')}
          </div>
        </div>
        <div className="dj-node__body">
          <b>{capitalize(data.type)}</b>:{' '}
          <a href={`/nodes/${data.name}`}>
            {data.type === 'source' ? data.table : data.display_name}
          </a>
          <Collapse
            collapsed={true}
            text={data.type !== 'metric' ? 'columns' : 'dimensions'}
          >
            <div className="dj-node__metadata">
              {data.type !== 'metric'
                ? columnsRenderer(data)
                : dimensionsRenderer(dimensions)}
            </div>
          </Collapse>
        </div>
      </div>
      <Handle
        type="source"
        position={Position.Right}
        style={{ backgroundColor: '#ccc' }}
      />
    </>
  );
}

export default memo(DJNode);
