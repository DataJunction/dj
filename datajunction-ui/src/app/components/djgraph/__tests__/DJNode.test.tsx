import * as React from 'react';
import { createRenderer } from 'react-test-renderer/shallow';

import { DJNode } from '../DJNode';

const renderer = createRenderer();

describe('<DJNode />', () => {
  it('should render and match the snapshot', () => {
    renderer.render(
      <DJNode
        id="1"
        data={{
          name: 'shared.dimensions.accounts',
          column_names: ['a'],
          type: 'source',
          primary_key: ['id'],
        }}
      />,
    );
    const renderedOutput = renderer.getRenderOutput();
    expect(renderedOutput).toMatchSnapshot();
  });
});
