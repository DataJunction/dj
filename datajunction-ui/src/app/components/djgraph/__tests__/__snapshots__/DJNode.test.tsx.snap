// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DJNode /> should render and match the snapshot 1`] = `
<React.Fragment>
  <Memo(Handle)
    position="left"
    style={
      Object {
        "backgroundColor": "#ccc",
      }
    }
    type="target"
  />
  <div
    className="dj-node__full"
    style={
      Object {
        "backgroundColor": "#7EB46150",
        "color": "#7EB461",
      }
    }
  >
    <div
      className="dj-node__header"
    >
      <div
        className="serif"
      >
        shared â–¶ dimensions
      </div>
    </div>
    <div
      className="dj-node__body"
    >
      <b>
        Source
      </b>
      :
       
      <a
        href="/nodes/shared.dimensions.accounts"
      />
      <Collapse
        collapsed={true}
        text="columns"
      >
        <div
          className="dj-node__metadata"
        >
          <tr>
            <td>
              <React.Fragment />
            </td>
            <td
              style={
                Object {
                  "textAlign": "right",
                }
              }
            />
          </tr>
        </div>
      </Collapse>
    </div>
  </div>
  <Memo(Handle)
    position="right"
    style={
      Object {
        "backgroundColor": "#ccc",
      }
    }
    type="source"
  />
</React.Fragment>
`;
